### PR TITLE
feat(flow): shared in-memory state for dataflow operator 

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#3187 
## What's changed and what's your intention?
A shared state of key-value pair for various state
in dataflow execution, called `Arrangement`

i.e: Mfp operator with temporal filter need to store it's future output so that it can add now, and delete later. To get all needed updates in a time span, use `get_updates_in_range`

And reduce operator need full state of it's output, so that it can query(and modify by calling `apply_updates`)
existing state, also need a way to expire keys. To get a key's current value, use `get` with time being `now`
so it's like:
 `mfp operator -> arrange(store futures only, no expire) -> reduce operator <-> arrange(full, with key expiring time) -> output`

Note the two way arrow between reduce operator and arrange, it's because reduce operator need to query existing state
 and also need to update existing state.

This state will be used in building&execution dataflow graph

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/flow/src/repr.rs [hunk 1]
```diff
@@ -45,11 +45,20 @@ pub type Timestamp = i64;
...
 
+pub type KeyValDiffRow = ((Row, Row), Timestamp, Diff);
+
 /// Convert a value that is or can be converted to Datetime to internal timestamp
 pub fn value_to_internal_ts(value: Value) -> Result<Timestamp, EvalError> {
+    let is_supported_time_type = |arg: &Value| {
+        let ty = arg.data_type();
+        matches!(ty, ConcreteDataType::Date(..))
+            | matches!(ty, ConcreteDataType::DateTime(..))
+            | matches!(ty, ConcreteDataType::Timestamp(..))
```
Comment (zhongzc) [inline src/flow/src/repr.rs:56]:
```suggestion
        matches!(ty, ConcreteDataType::Date(_) | ConcreteDataType::DateTime(_) | ConcreteDataType::Timestamp(_))
```

File: src/flow/src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+
+/// Determine when should a key expire according to it's event timestamp in key
+///
+/// TODO(discord9): find a better way to handle key expiration, like write to disk or something instead of throw away
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Deserialize, Serialize)]
+pub struct KeyExpiryManager {
+    /// a map from event timestamp to key, used for expire keys
+    event_ts_to_key: BTreeMap<Timestamp, BTreeSet<Row>>,
+    /// duration after which a key is considered expired, and will be removed from state
+    key_expiration_duration: Option<Timestamp>,
```
Comment (zhongzc) [inline src/flow/src/utils.rs:35]: This type should not be `Timestamp`

Comment (zhongzc) [reply src/flow/src/utils.rs:35]: how about naming it `*ttl`

File: src/flow/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        if !self.is_written {
+            self.is_written = true;
+        }
+        for ((key, val), ts, diff) in updates {
+            // keep rows with expired event timestamp from being updated
+            if let Some(s) = &mut self.expire_state {
+                if s.is_key_expired(now, &key)? {
+                    continue;
+                }
+                s.update_event_ts(now, &key)?;
```
Comment (zhongzc) [inline src/flow/src/utils.rs:152]: Duplicate calls... `is_key_expried`x2 and `extract_event_ts`x3

File: src/flow/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+        range: R,
+    ) -> Vec<KeyValDiffRow> {
+        let mut result = vec![];
+        for (_ts, batch) in self.spine.range(range) {
+            for (key, updates) in batch.clone() {
+                for (val, ts, diff) in updates {
+                    result.push(((key.clone(), val), ts, diff));
+                }
+            }
+        }
```
Comment (zhongzc) [inline src/flow/src/utils.rs:223]: Can we eliminate some `clone`s here?

File: src/flow/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+            let expire_time = if let Some(t) = s.get_expire_time(now) {
+                t
+            } else {
+                // never expire
+                return;
+            };
+            // find all keys smaller than or equal expire_time and silently remove them
+            let mut after = s.event_ts_to_key.split_off(&(expire_time + 1));
+            std::mem::swap(&mut s.event_ts_to_key, &mut after);
+            let before = after;
```
Comment (zhongzc) [inline src/flow/src/utils.rs:239]: It feels like a bit of abstract leakage

Comment (discord9) [reply src/flow/src/utils.rs:239]: > It feels like a bit of abstract leakage

If you are refering to using `drain_range`, sadly there are cavets related to lifetime and soundness, so no avaliable functions now: https://internals.rust-lang.org/t/pre-rfc-btreemap-btreeset-drain-with-a-range/15911 maybe I can wrap `split_off` into a simple handwritten drain_range that `drain_range(input_map)->(drain_result_map, drain_rest_of_map)`? On a second thought, that might be a good idea

File: src/flow/src/utils.rs (hunk 2 above)
Comment (discord9) [reply src/flow/src/utils.rs:35]: > how about naming it `*ttl`

Rename it to `Duration` instead, but it's still a newtype of i64(`type Duration = i64;`)

File: src/flow/src/utils.rs (hunk 4 above)
Comment (discord9) [reply src/flow/src/utils.rs:223]: > Can we eliminate some `clone`s here?

The second condition need to return a local variable as get's result(it's like a multiversion thing that get value at a given time) so I'm afraid I have to clone in here

File: src/flow/src/repr.rs [hunk 6]
```diff
@@ -42,14 +42,29 @@ pub type Diff = i64;
 /// System-wide default timestamp type
 pub type Timestamp = i64;
 
+/// System-wide default duration type
+pub type Duration = i64;
```
Comment (waynexia) [inline src/flow/src/repr.rs:46]: Maybe also document the precision (millisecond)

File: src/flow/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+            .as_ref()
+            .map(|e| e.eval(&row.inner))
+            .transpose()?
+            .map(value_to_internal_ts)
+            .transpose()?;
+        Ok(ts)
+    }
+
+    pub fn get_expire_time(&self, now: Timestamp) -> Option<Timestamp> {
+        self.key_expiration_duration.map(|d| now - d)
```
Comment (waynexia) [inline src/flow/src/utils.rs:54]: Should this be `now + d`?

File: src/flow/src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+            .insert(row.clone());
+        Ok(true)
+    }
+}
+
+/// A shared state of key-value pair for various state
+/// in dataflow execution
+///
+/// i.e: Mfp operator with temporal filter need to store it's future output so that it can add now, and delete later.
+/// To get all needed updates in a time span, use `get_updates_in_range`
```
Comment (waynexia) [inline src/flow/src/utils.rs:87]: Sugg.: enclose those tokens in `[]` and there will be a link on generated docs. E.g.:
```suggestion
/// To get all needed updates in a time span, use [`get_updates_in_range`]
```

File: src/flow/src/utils.rs (hunk 7 above)
Comment (discord9) [reply src/flow/src/utils.rs:54]: > Should this be `now + d`?

Bad naming, it's actually geting timestamp that should be expired by now, renamed to `compute_expiration_timestamp` instead.

File: src/flow/src/utils.rs (hunk 4 above)
Comment (zhongzc) [reply src/flow/src/utils.rs:215]: get it

File: src/flow/src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+                        final_val = compact_diff_row(final_val, new_row.clone());
+                    }
+                }
+            }
+            final_val
+        }
+    }
+}
+
+fn compact_diff_row(old_row: Option<DiffRow>, new_row: DiffRow) -> Option<DiffRow> {
```
Comment (zhongzc) [inline src/flow/src/utils.rs:267]:
```suggestion
fn compact_diff_row(old_row: Option<DiffRow>, new_row: &DiffRow) -> Option<DiffRow> {
```

File: src/flow/src/utils.rs [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+                // never expire
+                return;
+            };
+            // find all keys smaller than or equal expire_time and silently remove them
+            let mut after = s.event_ts_to_key.split_off(&(expire_time + 1));
+            std::mem::swap(&mut s.event_ts_to_key, &mut after);
+            let before = after;
+            for (_, keys) in before.into_iter() {
+                for key in keys {
+                    self.spine.first_entry().map(|e| e.into_mut().remove(&key));
```
Comment (zhongzc) [inline src/flow/src/utils.rs:234]: why only remove the oldest entry?

File: src/flow/src/utils.rs [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+                    compacted_batch.insert(key, smallvec![compacted_update]);
+                } else {
+                    compacted_batch.remove(&key);
+                }
+            }
+        }
+
+        // insert the compacted batch into spine with key being `now`
+        let mut first_entry = BTreeMap::from([(now, compacted_batch)]);
+        self.spine.append(&mut first_entry);
```
Comment (zhongzc) [inline src/flow/src/utils.rs:199]:
```suggestion
        self.spine.insert(now, compacted_batch);
```

File: src/flow/src/utils.rs [hunk 12]
```diff
@@ -264,19 +263,19 @@ impl Arrangement {
...
     match (old_row, diff) {
-        (Some((row, _old_ts, old_diff)), diff) if row == val && old_diff + diff == 0 => {
+        (Some((row, _old_ts, old_diff)), diff) if row == *val && old_diff + diff == 0 => {
             // the key is deleted now
             None
         }
-        (Some((row, _old_ts, old_diff)), diff) if row == val && old_diff + diff != 0 => {
-            Some((val, ts, old_diff + diff))
+        (Some((row, _old_ts, old_diff)), diff) if row == *val && old_diff + diff != 0 => {
+            Some((val.clone(), *ts, old_diff + *diff))
```
Comment (zhongzc) [inline src/flow/src/utils.rs:274]:
```suggestion
            Some((row, *ts, old_diff + *diff))
```

**Review by zhongzc — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
LGTM

**Pull request merged**